### DISCUSSION TITLE
New .NET 7 Regex Source Generator: Syntax too obtrusive

### DISCUSSION THREAD

**Discussion by SetTrend** [dotnet/csharplang#6509] [General] [answered]
The new feature is nice, but I believe it suffers from a [very obtrusive syntax](https://devblogs.microsoft.com/dotnet/announcing-dotnet-7-preview-2/#after).

Couldn't you make an attribute or keyword alone (e.g. `const`) do the job at compile time to switch the Regular Expression to the new Regex Source Generator?

**Comment by huoyaoyuan** [comment, marked as answer]
The generator belongs to dotnet/runtime repo.
The syntax looks necessary to me. 

> Couldn't you make an attribute or keyword alone (e.g. const) do the job at compile time to switch the Regular Expression to the new Regex Source Generator?

It can hardly be "switching". The code structure of using source generator and using constructor are fundamentally different. A method (or partial property, if it exists) is necessary to wrap the details.

**Reply by SetTrend** [reply]
The MS blog directed here.

> The code structure of using source generator and using constructor are fundamentally different. A method (or partial property, if it exists) is necessary to wrap the details.

Yes, I comprehend. Yet, this nitty gritty detail could be cloaked by the compiler.

I propose not to burden the programmer with the currently suggested verbose syntax.

It's just like with the `await`/`async` C# keywords. They're cloaking a whole bunch of code lines that would otherwise have to be programmed by each and every programmer over and over again.

**Reply by HaloFour** [reply]
The C# compiler doesn't know anything about how this particular source generator, or any particular source generator works.  Having to change the language in order to support a specific source generator goes against the reason source generators exist.

**Reply by SetTrend** [reply]
I see. Thank you for enlightening me.

I'm not too deep in the technical details of how this is accomplished. Please pardon me for that.

Yet, something _is_ happening at compile time, right? Something _is_ triggering the usage of the alternative RegEx source generator at compile time. If it's not the _compiler_ doing that … what's doing this?

**Reply by HaloFour** [reply]
The compiler does trigger the source generator.  Source generators can then generate additional C# source which is added to the compilation, but they cannot modify any existing source.  This generator keys off of an existing `partial` method that has the `RegexGenerator` attribute and uses that to emit an implementation to the `partial` method, which the compiler then combines into a single actual method.

**Reply by jnm2** [reply]
You can also view the generated partial class that adds the functionality within the Analyzers node of the project in Solution Explorer, to see what is necessary to generate to get this functionality to work.

**Reply by 333fred** [reply]
@SetTrend to be more explicit, source generators are _user_ authored code. The user in this case happens to be the standard library, but you could write your own generator using the same system and the compiler would know as much about it as it knows about the regex generator.

**Reply by SetTrend** [reply]
Thanks to all of you for clarifying this. I just read through the Roslyn Source Generators documentation. I now see it's a more general approach, and compiled Regular Expressions is rather one implementation of such Source Generator.

**Reply by jnm2** [reply]
Though, one could imagine inventing a syntax compilers could generally use to do this sort of thing  inline without requiring a partial method declaration, with full generality and applications outside regex. It sounds like a hard sell.

**Comment by RenderMichael** [comment]
I think it could be improved with the introduction of partial properties. since the method call might give the impression that there's computation going on.
